orb-ui/cone/src/button.rs

```diff
@@ -1,86 +1,90 @@
+    /// Used to signal that the button's task should be cleanly terminated.
+    pub kill_tx: oneshot::Sender<()>,
+}
+
+impl PartialEq for ButtonState {
+    fn eq(&self, other: &Self) -> bool {
+        matches!(
+            (self, other),
+            (ButtonState::Pressed, ButtonState::Pressed)
+                | (ButtonState::Released, ButtonState::Released)
+        )
+    }
 }
 
 /// Poll the button state.
 /// Events are sent to the event queue when the button is pressed or released
-/// The thread that polls the button is also used to check the connection status
 impl Button {
     pub(crate) fn spawn(
-        event_queue: mpsc::UnboundedSender<ConeEvents>,
-        connection: Arc<Mutex<Status>>,
-    ) -> eyre::Result<Self> {
-        let mut device: Ft4232h = Ftdi::with_index(7)?.try_into()?;
-        let mask: u8 = !BUTTON_GPIO_MASK; // button pin as input, all others as output
-        device.set_bit_mode(mask, BitMode::AsyncBitbang)?;
-        debug!("Button GPIO initialized");
+        event_queue: broadcast::Sender<ConeEvent>,
+    ) -> eyre::Result<(Self, ButtonJoinHandle)> {
+        let mut device: Ft4232h =
+            Ftdi::with_index(CONE_FTDI_BUTTON_INDEX)?.try_into()?;
+        device.set_bit_mode(BUTTON_GPIO_DIRECTION, BitMode::AsyncBitbang)?;
+        tracing::debug!("Button GPIO initialized");
 
-        let terminate = Arc::new(AtomicBool::new(false));
-        let terminate_clone = Arc::clone(&terminate);
+        let (kill_tx, mut kill_rx) = oneshot::channel();
 
         // spawn a thread to poll the button
-        let thread_handle = std::thread::spawn(move || {
+        let thread_handle = tokio::task::spawn(async move {
             // keep state so that we send an event only on state change
-            let mut last_state = false;
+            let mut last_state = ButtonState::Released;
             loop {
-                if terminate_clone.load(Ordering::Relaxed) {
-                    return;
-                }
-                match device.bit_mode() {
-                    Ok(mode) => {
-                        // connected
-                        if let Ok(mut status) = connection.lock() {
-                            *status = Status::Connected;
-                        }
-                        // button is active low
-                        let pressed = mode & BUTTON_GPIO_MASK == 0;
-                        if pressed != last_state {
-                            if let Err(e) =
-                                event_queue.send(ConeEvents::ButtonPressed(pressed))
-                            {
-                                tracing::error!("Error sending event: {:?} - no receiver? stopping producer", e);
-                                return;
+                select! {
+                    _ = &mut kill_rx => return Ok(()),
+                    _ = tokio::time::sleep(std::time::Duration::from_millis(BUTTON_POLL_INTERVAL_MS)) => {
+                        match device.bit_mode() {
```

Comment:
you're blocking in async. use tokio::task::spawn_blocking instead of spawn(). The first attempt used std::thread which is why I never complained about that part :p